### PR TITLE
Enabling general search for ports, devices, and more

### PR DESCRIPTION
**Pull request by hachpai**
Three years ago, I [suggested](https://community.librenms.org/t/port-search-query-based-on-ifalias-using-global-search/1356) a feature in the API but I was lacking of time and experience to write it. Now it's done. I exposed the generic search (used by ajax search bar) to the API.
For now, i just did it with ports (because it's really what I need). But it's trivial to extend it.
It's my quick contribution and almost my first time coding in PHP so you should expect a WIP, but working. Here's a curl example: `curl -H "X-Auth-Token: XXXX" 'http://127.0.0.1:8000/api/v0/search_ports?search=[what you want]'`

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/port-search-query-based-on-ifalias-using-global-search/1356/9

**Comment by murrant** [comment]
Nice, but you are re-arranging and pulling in a ton of complex code for something that only requires a single query.

```php
$value = "%$search%";
\App\Models\Port::hasAccess(Auth::user())
    ->orderBy('ifName')
    ->where('ifAlias', 'like', $value)
    ->orWhere('ifDescr', 'like', $value)
    ->orWhere('ifName', 'like', $value)
    ->get();
```

**Comment by hachpai** [comment]
You're right. I tried to stick to the logic without diving deep :-)

**Comment by hachpai** [comment]
By user-pending, you expect me to change for the implementation suggested by @murrant ?
Or we go for a more generic search and expose the same logic than ajax search?
It's more a kind of a multi purpose search engine and we cover all the use case in one pull.

**Comment by murrant** [comment]
Yes, I suggest keeping the scope of this PR just for ports.

**Comment by hachpai** [comment]
Just tested and pushed the new version. Is it ok for you?

File: includes/html/api_functions.inc.php [hunk 1]
```diff
@@ -72,6 +72,19 @@ function api_not_found()
     return api_error(404, "This API route doesn't exist.");
 }
 
+function search_ports(\Illuminate\Http\Request $request)
+{
+    $search = $request->get('search');
```
Comment (SourceDoctor) [inline includes/html/api_functions.inc.php:77]: what happens if search is empty or an empty string?

Comment (murrant) [reply includes/html/api_functions.inc.php:77]: it searchs for %%

Comment (SourceDoctor) [reply includes/html/api_functions.inc.php:77]: doesn't that mean "no filter", so all ports are searched?
not sure if it's such good on thousands of ports which will be returned as result

**Comment by murrant** [comment]
This pull request has been mentioned on **LibreNMS Community**. There might be relevant details there:

https://community.librenms.org/t/v1-64-release-changelog-may-2020/12290/1

**Pull request merged**